rfcs/rfc-4-text-selection-design.md

```diff
@@ -0,0 +1,906 @@
+
+This RFC came from the need to render a text cursor, but it actually isn't about that. The primary objective of this RFC is to set the foundation for all (or at least *most*) things related to text selection, which rendering a controllable caret falls into. This document should be an outline for implementation, not the implementation itself. The real implementation details should be discussed in a separate RFC or directly in their own PR, using this RFC as a base.
+
+> This RFC actually *does* cover some implementation details such as traversing the node tree and algorithms for calculating the position of a character. These were mainly included since they're so vital to all other systems. How we do some of these things might affect how we decide to implement other features (for example, how we determine the selected line helps inform how the cursor should move up or down).
+
+Below are the goals for the *entire* text-selection feature and include what this document will be outlining for the most part.
+
+**Main Goals**
+
+* Visualize a cursor within focused text fields
+  * Allow it to be placed and moved
+
+* Allow text within widgets to be selectable (such as the contents of a `TextBox`)
+  * Visually indicate the selection
+
+
+**Secondary Goals**
+
+* Allow text across widgets to be selectable
+
+**Tertiary Goals**
+
+* All of the above with future accessibility in mind
+
+## Additional Context
+
+### Resources
+
+A lot of the content of this RFC is taken from the DOM Standard [specs](https://dom.spec.whatwg.org/). Of course, it's been paraphrased and translated to fit the needs of Kayak. However, it's still great resource for understanding how the web has chosen to handle things.
+
+### Discussion
+
+Parts of this RFC require discussion before a consensus is reached on how to tackle a certain problem. Anything is up for discussion really, but the parts that really need it are marked with the `💬` emoji.
+
+## Terminology
+
+### Character
+
+Unless otherwise stated, a <em>**character**</em> is a blanket term referring to a *variable-width* character. This means that a character is not necessarily a single byte and could be made up of one or more, as is often the case when working with encodings such as [UTF-8](https://en.wikipedia.org/wiki/UTF-8). 
+
+It's important to account for this because it better prepares us for multi-language support.
+
+Additionally, when it comes to text selection, we generally want the [*grapheme cluster*](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries). This is what—to a user— is typically viewed as a "character".
+
+> Rust's built-in [`chars()`](https://doc.rust-lang.org/stable/std/primitive.str.html#method.chars) method breaks strings down into *Unicode Scalar Values*. For segmentation along grapheme clusters, an external crate like [`unicode-segmentation`](https://crates.io/crates/unicode-segmentation) will likely need to be used.
```

Comment:
There a few different crates for managing this. It might be worth while looking into using some of those.